react/src/components/Chat/ChatCard.tsx

```diff
@@ -0,0 +1,308 @@
+
+  const { endpoint, setEndpoint } = useEndpoint(selectedEndpoint);
+  const { models, modelId, setModelId } = useModels(
+    chat.provider,
+    fetchKey,
+    endpoint,
+  );
+  const { agents, agent, setAgent } = useAgents(chat.provider);
+  const [sync, setSync] = useState(chat.sync);
+
+  const baseURL = endpoint?.url
+    ? new URL(chat.provider.basePath, endpoint?.url ?? undefined).toString()
+    : undefined;
+
+  const allowCustomModel = isEmpty(models);
+  const providerSettings = {
+    baseURL: allowCustomModel
+      ? formRef.current?.getFieldValue('baseURL')
+      : baseURL,
+    modelId: allowCustomModel
+      ? formRef.current?.getFieldValue('modelId')
+      : modelId,
+    apiKey: allowCustomModel
+      ? formRef.current?.getFieldValue('token')
+      : chat.provider.apiKey,
+  };
+
+  const {
+    error,
+    messages,
+    input,
+    setInput,
+    stop,
+    isLoading,
+    append,
+    setMessages,
+  } = useChat({
+    id: `${chat.id}-${endpoint?.endpoint_id}-${agent?.id ?? 'none'}-${modelId}`,
+    api: baseURL,
+    credentials: chat.provider.credentials,
+    body: {
+      modelId: modelId,
+    },
+    fetch: async (input, init) => {
+      if (fetchOnClient || modelId === 'custom') {
+        const body = JSON.parse(init?.body as string);
+        const provider = createOpenAI({
+          baseURL: providerSettings.baseURL,
+          apiKey: providerSettings.apiKey || 'dummy',
+        });
+        const result = streamText({
+          abortSignal: init?.signal || undefined,
+          model: wrapLanguageModel({
+            model: provider(providerSettings.modelId),
+            middleware: extractReasoningMiddleware({ tagName: 'think' }),
+          }),
+          messages: body?.messages,
+          system: agent ? (agent.config.system_prompt ?? '') : '',
+        });
+
+        setStartTime(Date.now());
+
+        return result.toDataStreamResponse({
+          sendReasoning: true,
+        });
+      } else {
+        return fetch(input, init);
+      }
+    },
+  });
+
+  return (
+    <Card
+      bordered
+      style={style}
+      styles={{ ...styles }}
+      title={
+        <ChatHeader
+          chat={chat}
+          allowCustomModel={allowCustomModel}
+          models={models}
+          modelId={modelId}
+          setModelId={setModelId}
+          endpoint={endpoint}
+          setEndpoint={setEndpoint}
+          agents={agents}
+          agent={agent}
+          setAgent={setAgent}
+          sync={sync}
+          setSync={setSync}
+          fetchKey={fetchKey}
+          closable={closable}
+          onCreateNewChat={onCreateNewChat}
+          onRequestClose={onRequestClose}
+          setMessages={setMessages}
+        />
+      }
+    >
+      <CustomModelForm
+        modelId={modelId}
+        baseURL={baseURL}
+        formRef={formRef}
+        allowCustomModel={allowCustomModel}
+        alert={formRef && <CustomModelAlert onClick={() => updateFetchKey()} />}
```

Comment:
When users click "Refresh", it should request the baseURL of the CustomModelForm.